scheds/rust/scx_tickless/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,696 @@
+
+/*
+ * Task @p is about to start running on a CPU.
+ */
+void BPF_STRUCT_OPS(tickless_running, struct task_struct *p)
+{
+	s32 cpu = scx_bpf_task_cpu(p);
+	struct cpu_ctx *cctx;
+	struct task_ctx *tctx;
+
+	cctx = try_lookup_cpu_ctx(cpu);
+	if (!cctx)
+		return;
+	cctx->started_at = scx_bpf_now() ? : 1;
+
+	tctx = try_lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Update the run timestamp (used to evaluate the used time slice).
+	 */
+	tctx->last_run_at = scx_bpf_now();
+
+	/*
+	 * Update the global vruntime as a new task is starting to use a
+	 * CPU.
+	 */
+	if (time_before(vtime_now, tctx->deadline))
+		vtime_now = tctx->deadline;
+}
+
+/*
+ * Task @p is about to release the CPU.
+ */
+void BPF_STRUCT_OPS(tickless_stopping, struct task_struct *p, bool runnable)
+{
+	s32 cpu = scx_bpf_task_cpu(p);
+	struct cpu_ctx *cctx;
+	struct task_ctx *tctx;
+	u64 slice;
+
+	cctx = try_lookup_cpu_ctx(cpu);
+	if (!cctx)
+		return;
+	cctx->started_at = 0;
+
+	tctx = try_lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Evaluate the time slice used by the task.
+	 */
+	slice = scx_bpf_now() - tctx->last_run_at;
+
+	/*
+	 * Update task's execution time (exec_runtime), but never account
+	 * more than 1s of runtime to prevent excessive de-prioritization
+	 * of CPU-intensive tasks (which could lead to starvation).
+	 */
+	if (tctx->exec_runtime < NSEC_PER_SEC)
+		tctx->exec_runtime += slice;
+
+	/*
+	 * Update task's vruntime.
+	 */
+	tctx->deadline += scale_inverse_fair(p, slice);
+}
+
+/*
+ * Task @p is entering the BPF scheduler.
+ */
+void BPF_STRUCT_OPS(tickless_enable, struct task_struct *p)
+{
+	struct task_ctx *tctx;
+
+	tctx = try_lookup_task_ctx(p);
+	if (!tctx)
+		return;
+
+	/*
+	 * Initialize the task vruntime to the current global vruntime.
+	 */
+	tctx->deadline = vtime_now;
+}
+
+/*
+ * Task @p is created.
+ */
+s32 BPF_STRUCT_OPS(tickless_init_task, struct task_struct *p,
+		   struct scx_init_task_args *args)
+{
+	struct task_ctx *tctx;
+
+	tctx = bpf_task_storage_get(&task_ctx_stor, p, 0,
+				    BPF_LOCAL_STORAGE_GET_F_CREATE);
+	if (!tctx)
+		return -ENOMEM;
+
+	return 0;
+}
+
+static int sched_timerfn(void *map, int *key, struct bpf_timer *timer)
+{
+	const struct cpumask *online_cpumask;
+	struct cpu_ctx *cctx;
+	u64 now = scx_bpf_now();
+	bool is_primary;
+	s32 cpu;
+
+	bpf_rcu_read_lock();
+	is_primary = is_primary_cpu(bpf_get_smp_processor_id());
+	if (!is_primary) {
+		scx_bpf_error("Scheduling timer executed on a non-primary CPU");
+		goto out_unlock;
+	}
+
+	online_cpumask = scx_bpf_get_online_cpumask();
+
+	bpf_for(cpu, 0, nr_cpu_ids) {
+		if (!bpf_cpumask_test_cpu(cpu, online_cpumask) || is_primary_cpu(cpu))
+			continue;
+
+		cctx = try_lookup_cpu_ctx(cpu);
+		if (!cctx)
+			continue;
+
+		if (time_before(now, cctx->started_at + slice_ns))
+			continue;
+
+		if (!scx_bpf_dsq_nr_queued(SCX_DSQ_LOCAL_ON | cpu) &&
+		    !scx_bpf_dsq_nr_queued(SHARED_DSQ))
+			continue;
+
+		scx_bpf_kick_cpu(cpu, SCX_KICK_PREEMPT);
```

Comment:
I *think* it should be possible to call `dispatch_all_cpus()` directly from the timer.

Reply:
Almost... I can't check the available dispatch slots from the timer:
```
swapper/0[0] triggered exit kind 1024:
  runtime error (kfunc with mask 0x2 called from an operation only allowing 0x0)

Backtrace:
  scx_bpf_dispatch_nr_slots+0x52/0x70
  bpf_prog_93597a2c6163323b_dispatch_all_cpus+0x1b7/0x224
  bpf_prog_4cfe5fb148bda16b_sched_timerfn+0xd1/0x28e
  bpf_timer_cb+0x6b/0xc0
```

Reply:
You don't need to check nr_slots from other paths. Dispatches don't go through the dispatch buffer outside of ops.dispatch().

Reply:
yep, I'll investigate more and see if we can move the dispatches in the timer handler